lib/menubar/menu_gen.lua

```diff
@@ -90,34 +90,38 @@ function menu_gen.generate()
 
     local result = {}
 
+    local entries = {}
     for _, dir in ipairs(menu_gen.all_menu_dirs) do
-        local entries = utils.parse_dir(dir)
-        for _, program in ipairs(entries) do
-            -- Check whether to include program in the menu
-            if program.show and program.Name and program.cmdline then
-                local target_category = nil
-                -- Check if the program falls at least to one of the
-                -- usable categories. Set target_category to be the id
-                -- of the first category it finds.
-                if program.categories then
-                    for _, category in pairs(program.categories) do
-                        local cat_key, cat_use =
-                            get_category_name_and_usage_by_type(category)
-                        if cat_key and cat_use then
-                            target_category = cat_key
-                            break
-                        end
+        for _, entry in ipairs(utils.parse_dir(dir)) do
+            entries[entry.Exec .. '\0' .. entry.Name] = entry
```

Comment:
This key is unused, this is just so that you get a unique key, right?
I'd propose to use the following then instead:

    for _, dir in ipairs(menu_gen.all_menu_dirs) do
        awful.util.table.merge(entries, utils.parse_dir(dir))
    end

(This would also preserve "some" order of the elements)